pkg/util/active_groups.go

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package util
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/grafana/dskit/services"
+	"go.uber.org/atomic"
+)
+
+type ActiveGroups struct {
+	mu                sync.RWMutex
+	timestampsPerUser map[string]map[string]*atomic.Int64 // map[user][group] -> timestamp
+	maxGroupsPerUser  int
+}
+
+func NewActiveGroups(maxGroupsPerUser int) *ActiveGroups {
+	return &ActiveGroups{
+		timestampsPerUser: map[string]map[string]*atomic.Int64{},
+		maxGroupsPerUser:  maxGroupsPerUser,
+	}
+}
+
+func (ag *ActiveGroups) UpdateGroupTimestampForUser(userID, group string, ts int64) {
+	ag.mu.RLock()
+	groupTimestamps := ag.timestampsPerUser[userID]
+	ag.mu.RUnlock()
+
+	if groupTimestamps == nil {
+		ag.mu.Lock()
+		if ag.timestampsPerUser[userID] == nil {
+			ag.timestampsPerUser[userID] = map[string]*atomic.Int64{group: atomic.NewInt64(ts)}
+		}
+		ag.mu.Unlock()
+		return
+	}
+
+	ag.mu.RLock()
+	groupTs := ag.timestampsPerUser[userID][group]
```

Comment:
Locking again is expensive, can we retrieve this before when we got the RMutex for the first time if we see that `groupTimetamps` is not nil?